### ISSUE TITLE
Support for validating resources before saving to local db

### ISSUE THREAD

**Issue by LZRS** [opensrp/fhircore#2853]
**Describe the feature request.**
 Add support for validating FHIR resources before getting saved to local DB to ensure that no invalid resources are saved. This would also help ensure that when sync to server, requests do not fail because of resource being invalid

**Additional context**
If available add any other context or screenshots about the feature here.

**Acceptance criteria**
A check list of all things to verify once the implementation by the engineer is complete

**Area path**
A list of ordered steps in the app on usage of the feature to support anyone testing it e.g. Code reviewer, QA e.g.
1. Login to the app
2. Open _Navigation bar_
3. Click on _Children register_
4. Click on _Child profile_
5. Click on _Edit profile_ from menu

**Implementation plan (For Engineers)**
The plan for implementing the solution e.g. via a _description_ or a _check list_ for the various ordered tasks that will need to be completed.
_i.e. Describe how you intend to solve the problem_

**Comment by dubdabasoduba** [comment]
@pld I think this would be a good addition to the FHIR Core app too. We can make it configurable with an opt out capability.

**Comment by pld** [comment]
kinda agree! but.. what happens if they are invalid? once someones trying to store a resource we wouldn't want to not persist it, I guess in theory we could redo the extraction after the config has been updated, like you could mark the QR as extraction pending or something.

But I feel like the more impactful place for effort would be on validating structuremaps to ensure they are only producing valid resources as outputs

**Comment by ndegwamartin** [comment]
Suppose we had a variant of this that only makes this type of validation check mandatory for Debug mode. That way during development it is easier for engineers to catch such early on. This can be in place regardless of the conclusion we make for this ticket.

**Comment by dubdabasoduba** [comment]
@ndegwamartin I agree this is more for the development stage and not necessary for use in production. The assumption here is that all SMs and CQL would generate correctly formatted resources by the time an app goes live

**Comment by dubdabasoduba** [comment]
@LZRS how far are we with this? cc @ndegwamartin

**Comment by pld** [comment]
@LZRS can we show the user the failed validation directly on the screen in a way they can copy it? the users of debug will be testers on our and other teams so their workflow will be

1. submit form
2. get error
3. copy error data into new issue

I think your suggestion is brings up a very good point. If the tester hits a non-critical error, we don't want them to be unable to continue testing until that non-critical error is resolved. How about we show the error in a very big clear way so the tester knows to create an issue, but we continue with the submission so that they can continue testing the workflow.

**Comment by pld** [comment]
OK sounds good
